### PR TITLE
Bound products to table view

### PR DESCRIPTION
**Pull request by Rabee96**
There is still the delete alert I will add it in the next pull request

### CODE REVIEW

File: client/src/pages/Products/index.tsx [hunk 1]
```diff
@@ -1,7 +1,100 @@
...
+    deleteProductAPI(id)
+      .then(result => {
+        console.log(result);
+        return result.data;
+      })
+      .then(success => setDeletedItem(success))
+      .catch(error => {
+        console.log('delete error', error);
+      });
+  };
```
Comment (ysalem-dev-89) [inline client/src/pages/Products/index.tsx:45]: We should define these functions when we use them, no need to pass them as props.

File: client/src/pages/Products/index.tsx [hunk 2]
```diff
@@ -1,7 +1,100 @@
...
+              <TableProductRow
+                key={index}
+                product={product}
+                startEditMode={startEditMode}
+                deleteProduct={deleteProduct}
+              />
+            ))
+          ) : (
+            <h1>Loading...</h1>
+          )}
```
Comment (ysalem-dev-89) [inline client/src/pages/Products/index.tsx:93]: If the returned request has no products, the message will still be 'Loading...', which is not right.
I suggest you create a state for loading. and if the request is succeeded we should check if it has data or not.
if not we can display something like 'No products found'

**Review by ysalem-dev-89 — CHANGES_REQUESTED** [review]

File: client/src/pages/Products/index.tsx [hunk 3]
```diff
@@ -1,7 +1,100 @@
...
+  };
+  useEffect(() => {
+    getAllProductsAPI()
+      .then(result => {
+        console.log('result', result);
+        setProducts(result.data);
+      })
+      .catch(err => {
+        console.log(err);
+      });
```
Comment (ysalem-dev-89) [inline client/src/pages/Products/index.tsx:54]: Should show this error to the user.

File: client/src/pages/Products/index.tsx [hunk 4]
```diff
@@ -1,7 +1,100 @@
...
+    setEditMode({ editMode: !editMode, productId });
+  };
+
+  const update = (id: number, updatedProduct: Product) => {
+    updateProductAPI(id, updatedProduct)
+      .then(result => result.data)
+      .then(product => setUpdatedItem(product))
+      .catch(error => {
+        console.error('error', error);
+      });
```
Comment (ysalem-dev-89) [inline client/src/pages/Products/index.tsx:32]: User should be notified that the update operation is failed. I suggest to use a toast message.

**Review by ysalem-dev-89 — CHANGES_REQUESTED** [review]

File: client/src/api/products.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import axios from './axios';
+import { Product } from '../interfaces/ProductInterface';
+
+export const getAllProductsAPI = () => axios.get('/products');
+export const updateProductAPI = (id: number, newData: Product) =>
+  axios.put(`/products/${id}`, newData);
+export const deleteProductAPI = (id: number) => axios.put(`/products/${id}`);
```
Comment (AhmedSafi97) [inline client/src/api/products.ts:7]: should be `axios.delete`

File: client/src/pages/Products/index.tsx (hunk 4 above)
Comment (AhmedSafi97) [reply client/src/pages/Products/index.tsx:32]: @Rabee96 yep and please all other places where you make api calls and handle the errors with the right message

File: client/src/pages/Products/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (AhmedSafi97) [reply client/src/pages/Products/index.tsx:45]: @Rabee96 we must follow the same code style `async/await`

File: src/scripts/DataGenerator.ts [hunk 6]
```diff
@@ -80,23 +77,4 @@ export default class DataGenerator {
     }
     return transProdsArr;
   }
-
```
Comment (AhmedSafi97) [inline src/scripts/DataGenerator.ts:83]: I know I made this change, but please restore it

**Review by AhmedSafi97 — CHANGES_REQUESTED** [review]

**Comment by AhmedSafi97** [comment]
please resolve the comments and conflict

**Review by ysalem-dev-89 — DISMISSED** [review]

File: src/controllers/ProductController.ts [hunk 7]
```diff
@@ -87,7 +87,7 @@ export default class ProductController {
     try {
       const { id } = req.params;
       await ProductQuery.deleteProduct(Number(id));
-      res.json({
+      res.status(202).json({
```
Comment (AhmedSafi97) [inline src/controllers/ProductController.ts:90]: please change this to `res.sendStatus(204)` only

when you delete a product we don't return anything(no response body) and the right status for this is 204 no content

**Review by AhmedSafi97 — APPROVED** [review]

**Review by ysalem-dev-89 — APPROVED** [review]
Great work, I liked the filter feature in Offcanvas. nicely done!

**Review by AhmedSafi97 — APPROVED** [review]

**Pull request merged**